### PR TITLE
[XCOFF] Add compiler version to an auxiliary symbol table entry

### PR DESCRIPTION
**Pull request by stephenpeckham**
C_FILE symbols. To match the behavior of the assembler and the legacy compiler, this includes using the generic ".file" name for the C_FILE symbol and generating the actual file name in an auxiliary entry.

### CODE REVIEW

File: llvm/include/llvm/MC/MCAssembler.h
Comment (daltenty) [inline llvm/include/llvm/MC/MCAssembler.h:497]: 1. Is there are reason to have a return value here? We don't check it elsewhere
2. I think there's an extra copy here? Probably either you should: take the parameter as string ref and do the conversion here; or you should use std::move
3. I'm not sure I follow why we need the empty check to prevent modifying the string?

```suggestion
  void setCompilerVersion(std::string CompilerVers) {
      CompilerVersion = std::move(CompilerVers);
  }
```

File: llvm/lib/MC/MCObjectStreamer.cpp
Comment (daltenty) [inline llvm/lib/MC/MCObjectStreamer.cpp:906]: Prefer str(), data() isn't necessarily null terminated according to the docs.

nit: can we fix the typo in the parameter name here while were at it? It shows up in the docs  🙂

```suggestion
    getAssembler().setCompilerVersion(CompilerVersion.str());
```

File: llvm/test/CodeGen/PowerPC/aix-available-externally-linkage-fun.ll
Comment (stephenpeckham) [reply llvm/test/CodeGen/PowerPC/aix-available-externally-linkage-fun.ll:9]: I tried using the capture expression in some cases, but in some cases, the wrong line matched and set the captured expression incorrectly.

File: llvm/include/llvm/MC/MCAssembler.h
Comment (stephenpeckham) [reply llvm/include/llvm/MC/MCAssembler.h]: I left the empty check, even though it's not strictly necessary. There can be multiple file symbols, but only a single compiler version.

File: llvm/lib/MC/MCObjectStreamer.cpp
Comment (stephenpeckham) [reply llvm/lib/MC/MCObjectStreamer.cpp]: Typo fixed in https://github.com/llvm/llvm-project/pull/80703

**Review by daltenty — APPROVED** [review]
LGTM, thanks!

**Comment by jplehr** [comment]
This introduced a spelling mistake that broke some builds.
https://lab.llvm.org/buildbot/#/builders/193/builds/46220

**Comment by nico** [comment]
This seems to break tests, even after the spelling fix: http://45.33.8.238/linux/130117/step_12.txt

Does that look related?

**Comment by asl** [comment]
@stephenpeckham Apparently this was merged without waiting for buildkite results or ignoring its results. Linux build failed due to obvious spelling mistakes. Will you please next time wait until CI to finish?

Thanks!

**Comment by asl** [comment]
@nico How about revert?

File: llvm/test/CodeGen/PowerPC/aix-xcoff-reloc.ll
Comment (asl) [inline llvm/test/CodeGen/PowerPC/aix-xcoff-reloc.ll:178]: This should not be here.

Comment (stephenpeckham) [reply llvm/test/CodeGen/PowerPC/aix-xcoff-reloc.ll:178]: Probably "version 18.0.0git" should be removed.

**Comment by stephenpeckham** [comment]
The other failing testcase was added after my PR was created.  I didn't notice the failure that occured after the merge.

**Comment by stephenpeckham** [comment]
PR https://github.com/llvm/llvm-project/pull/80902 created.  Waiting for build results

**Comment by AtariDreams** [comment]
Okay my PR broke too but only for this build for some reason

**Pull request merged**
